services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,630 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import javax.security.auth.x500.X500Principal;
+
+import org.eclipse.hono.deviceregistry.mongodb.config.MongoDbBasedTenantsConfigProperties;
+import org.eclipse.hono.deviceregistry.mongodb.model.TenantDto;
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbCallExecutor;
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbDocumentBuilder;
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbServiceUtils;
+import org.eclipse.hono.deviceregistry.util.DeviceRegistryUtils;
+import org.eclipse.hono.deviceregistry.util.Versioned;
+import org.eclipse.hono.service.management.Id;
+import org.eclipse.hono.service.management.OperationResult;
+import org.eclipse.hono.service.management.Result;
+import org.eclipse.hono.service.management.tenant.Tenant;
+import org.eclipse.hono.service.management.tenant.TenantManagementService;
+import org.eclipse.hono.service.tenant.TenantService;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.TenantConstants;
+import org.eclipse.hono.util.TenantResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+
+import io.opentracing.Span;
+import io.vertx.core.AbstractVerticle;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.mongo.IndexOptions;
+import io.vertx.ext.mongo.MongoClient;
+import io.vertx.ext.mongo.MongoClientDeleteResult;
+
+/**
+ * An implementation of the tenant service and the tenant management service
+ * that uses a mongodb database to store tenants.
+ * <p>
+ * On startup this adapter tries to find the tenant collection, if not found, it gets created.
+ */
+@SuppressWarnings("RedundantTypeArguments")
+@Component
+@Qualifier("serviceImpl")
+@ConditionalOnProperty(name = "hono.app.type", havingValue = "mongodb")
+public final class MongoDbBasedTenantService extends AbstractVerticle implements TenantService, TenantManagementService {
+
+    private static final Logger log = LoggerFactory.getLogger(MongoDbBasedTenantService.class);
+
+    private MongoDbCallExecutor mongoDbCallExecutor;
+    private MongoClient mongoClient;
+    private MongoDbBasedTenantsConfigProperties config;
+
+    /**
+     * Sets an the {@code MongoDbCallExecutor} and {@code MongoClient} for the mongodb connection.
+     *
+     * @param mongoDbCallExecutor An instance of the configured, connection ready mongodb executor singleton.
+     * @throws NullPointerException if the mongoDbCallExecutor is {@code null}.
+     */
+    @Autowired
+    public void setExecutor(final MongoDbCallExecutor mongoDbCallExecutor) {
```

Comment:
For Line 78: Since it sets both MongoDbCallExecutor and MongoClient I found the description more suitable to describe what the method does.

Reply:
> For Line 78: Since it sets both MongoDbCallExecutor and MongoClient I found the description more suitable to describe what the method does.

Including the `mongoClient` is fine. `Sets an the {@code MongoDbCallExecutor} and {@code MongoClient} for the mongodb connection.` IMHO it isn't grammatically correct. How about something thing like `Sets an instance of the {@code MongoDbCallExecutor} and {@code MongoClient}.`

Reply:
done